plugin/stripcomments/stripcomments.go

```diff
@@ -0,0 +1,105 @@
+package stripcomments
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"slices"
+
+	"github.com/safedep/code/core"
+	"github.com/safedep/dry/log"
+	sitter "github.com/smacker/go-tree-sitter"
+)
+
+type StripCommentsCallback func(f core.File, r io.Reader) error
+
+type stripCommentsPlugin struct {
+	stripCommentsCallback StripCommentsCallback
+}
+
+// Verify contract
+var _ core.TreePlugin = (*stripCommentsPlugin)(nil)
+
+func NewStripCommentsPlugin(stripCommentsCallback StripCommentsCallback) *stripCommentsPlugin {
+	return &stripCommentsPlugin{
+		stripCommentsCallback: stripCommentsCallback,
+	}
+}
+
+func (p *stripCommentsPlugin) Name() string {
+	return "StripCommentsPlugin"
+}
+
+var supportedLanguages = []core.LanguageCode{core.LanguageCodePython, core.LanguageCodeJavascript}
+
+func (p *stripCommentsPlugin) SupportedLanguages() []core.LanguageCode {
+	return supportedLanguages
+}
+
+func (p *stripCommentsPlugin) AnalyzeTree(ctx context.Context, tree core.ParseTree) error {
+	lang, err := tree.Language()
+	if err != nil {
+		return fmt.Errorf("failed to get language: %w", err)
+	}
+
+	file, err := tree.File()
+	if err != nil {
+		return fmt.Errorf("failed to get file: %w", err)
+	}
+
+	treeData, err := tree.Data()
+	if err != nil {
+		return fmt.Errorf("failed to get tree data: %w", err)
+	}
+
+	log.Debugf("stripcomments - Analyzing tree for language: %s, file: %s\n",
+		lang.Meta().Code, file.Name())
+
+	// @TODO - If performance issues arise due to copying the entire file contents into memory,
+	// consider storing in a temporary file and returning a reader to that file
+	var output bytes.Buffer
+	stripComments(tree.Tree().RootNode(), *treeData, &output)
+
+	return p.stripCommentsCallback(file, &output)
+}
+
+func stripComments(node *sitter.Node, source []byte, output *bytes.Buffer) {
+	if node == nil {
+		return
+	}
+
+	// Skip comments and standalone doc strings
+	if node.Type() == "comment" || (node.Type() == "string" && isStandaloneDocstring(node)) {
+		return
+	}
+
+	// Preserve leading whitespace and newlines for non-comment nodes
+	start := node.StartByte()
+	end := node.EndByte()
+	if node.ChildCount() == 0 {
+		output.Write(source[start:end])
+	} else {
+		for i := 0; i < int(node.ChildCount()); i++ {
+			child := node.Child(i)
+			stripComments(child, source, output)
```

Comment:
I think we should explore if we can use `Iterator` instead of recursively calling function and deal with the complexities of recursion ourselves. See this:

https://pkg.go.dev/github.com/smacker/go-tree-sitter#Iterator

May be we can use a BFS / DFS iterator over the nodes

Reply:
The trailing/leading whitespaces and tabspaces are not available in the `node.Content(*source)` of sitter nodes. To preserve them (especially needed for python), it is essential to copy the source data between `StartByte` and `EndByte` of adjacent nodes as it is, into the stripped bytes.

Keeping it as recursive, to achieve this